[Site Isolation] Add support for text manipulation in cross-site frames

szewai opened the pull request:
#### ca8a33f183972bba6a96003dc5d6d29709c6490e
<pre>
[Site Isolation] Add support for text manipulation in cross-site frames
<a href="https://rdar.apple.com/150383885">rdar://150383885</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=292339">https://bugs.webkit.org/show_bug.cgi?id=292339</a>

Reviewed by Ryosuke Niwa.

To make text manipulation work with site isolation, UI process should send message to all frame processes and collect
replies from all of them. This patch also ensures web process inspects all local frames by replacing
traverseNextRendered() with traverseNext(), as traverseNextRendered() will stop early when encountering a remote frame.

API test: SiteIsolation.CompleteTextManipulation
          SiteIsolation.CompleteTextManipulationFailsInSomeFrame

* Source/WebCore/editing/TextManipulationController.cpp:
(WebCore::TextManipulationController::completeManipulation):
* Source/WebCore/editing/TextManipulationController.h:
* Source/WebCore/editing/TextManipulationControllerManipulationFailure.h:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _completeTextManipulation:completion:]):
(-[WKWebView _completeTextManipulationForItems:completion:]):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::startTextManipulations):
(WebKit::WebPageProxy::completeTextManipulation):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::startTextManipulations):
(WebKit::WebPage::completeTextManipulation):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(-[SiteIsolationTextManipulationDelegate init]):
(-[SiteIsolationTextManipulationDelegate _webView:didFindTextManipulationItems:]):
(-[SiteIsolationTextManipulationDelegate items]):
(TestWebKitAPI::createToken):
(TestWebKitAPI::createItem):
(TestWebKitAPI::TEST(SiteIsolation, CompleteTextManipulation)):
(TestWebKitAPI::(SiteIsolation, CompleteTextManipulationFailsInSomeFrame)):
(TestWebKitAPI::TEST(SiteIsolation, DISABLED_Events)): Deleted.

Canonical link: <a href="https://commits.webkit.org/294454@main">https://commits.webkit.org/294454@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/101152da7de89e220592f0be0441abf1e26afae1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/101824 "18 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/21492 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/11808 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/106982 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/52458 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/103864 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/21800 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/29999 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/77557 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34554 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/104831 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16844 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91946 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57868 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/101298 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/16671 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/9965 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/51809 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/86525 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10041 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/109339 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28957 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/21331 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/86509 "Found 1 new test failure: imported/w3c/web-platform-tests/service-workers/service-worker/redirected-response.https.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/29318 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/88147 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/86081 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21910 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30835 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/8556 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/23128 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/28885 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/34177 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/28696 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/32019 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/30255 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -14936,11 +14935,45 @@ void WebPageProxy::didFindTextManipulationItems(const Vector<WebCore::TextManipu
+        {
+            return adoptRef(*new TextManipulationCallbackAggregator(WTFMove(callback)));
+        }
+        ~TextManipulationCallbackAggregator()
```
Maybe insert a blank line between each function definition?

rniwa commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -14936,11 +14935,45 @@ void WebPageProxy::didFindTextManipulationItems(const Vector<WebCore::TextManipu
+
+            if (m_allFailed)
+                m_allFailed = false;
+            m_failures.appendVector(failures);
```
Hm... this doesn't seem right to me. If one of the process fails all of them, we'd get `allFailed` true but that doesn't necessarily mean `m_allFailed` is set to `true`. In that case, we need to generate errors for all the items which have thus far failed.

szewai replied:
Fixed by making web process also report the item indexes that have been successfully replaced, so that UI process can compute the state of each item and generate errors for the ones that are not successfully replaced (and don't have an explicit error).

rniwa commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -14934,13 +14933,70 @@ void WebPageProxy::didFindTextManipulationItems(const Vector<WebCore::TextManipu
+            ASSERT(RunLoop::isMain());
+            HashSet<unsigned> resultIndexes;
+            for (auto& failure : m_result.failures)
+                resultIndexes.add(failure.index + 1);
```
This whole business about adding 1 to avoid 0 is very confusing.
I think we should use `UnsignedWithZeroKeyHashTraits` instead.

rniwa replied:
If we used `BitVector` instead, we don't need to have this `HashSet` at all so that might be better altogether.

rniwa commented on Source/WebCore/editing/TextManipulationController.cpp:
```diff
@@ -717,8 +718,12 @@ auto TextManipulationController::completeManipulation(const Vector<WebCore::Text
+            continue;
+        }
+
+        succeededIndexes.append(i);
```
Instead of appending a full integer to store the index of success, why don't we create BitVector of length matching the length of completionItems? Then you can just check BitVector::get(index) to check whether the given index has succeeded or not.

szewai replied:
`BitVector` is not supported in IPC message yet. I can use `BitVector` here but will need to convert it to `Vector` for sending reply to UI process.

rniwa replied:
Oh I see. That's unfortunate.

rniwa commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
```diff
@@ -4369,6 +4402,214 @@ HTTPServer server({
+        { "/apple"_s, { iframeBytes } },
+    }, HTTPServer::Protocol::HttpsProxy);
+
+    bool receivedMessage1 = false;
```
Can we give them more descriptive variable names like `didLoad`, `didReceiveMainFrameContent`, and `didReceiveIframeContent`?

rniwa commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
```diff
@@ -4369,6 +4402,214 @@ HTTPServer server({
+        auto value2 = [NSNumber numberWithBool:item2.isSubframe];
+        return [value1 compare:value2];
+    };
+    RetainPtr sortedItems = [items.get() sortedArrayUsingComparator:sortFunction];
```
Why do we need to sort the results? Given we're waiting for both the main frame and the iframe to load, the order should be deterministic: prefix DFS.

szewai replied:
The items are received from different web processes, so there is no guarantee they will arrive UI process in order.
(i.e. UI process sends "start" to WP1, WP2, WP3 in order and the replies may not come in WP1, WP2, WP3 order); I've seen the iframe process replies faster when running test locally.

rniwa replied:
Hm... maybe we should check with the translation team to make sure that's okay. Right now, iframe content comes in the order they appear. Changing the order of it could affect translation results.

rniwa commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
```diff
@@ -4369,6 +4402,214 @@ HTTPServer server({
+            let element = document.getElementById('iframe');
+            element.parentNode.removeChild(element);
+        }
+        var messageCount = 0;
```
Maybe use `let` for consistency?

rniwa commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
```diff
@@ -4369,6 +4402,214 @@ HTTPServer server({
+    EXPECT_EQ(items.get().count, 3UL);
+    auto item1 = [sortedItems objectAtIndex:0];
+    auto item2 = [sortedItems objectAtIndex:1];
+    auto item3 = [sortedItems objectAtIndex:1];
```
Isn't this supposed to be at objectAtIndex:2? We should probably use a distinct content for frame 2 and frame 3 to avoid this kind of error.

rniwa commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
```diff
@@ -4369,6 +4402,214 @@ HTTPServer server({
+        }, false);
+    </script>
+    <iframe id='iframe' src='https://apple.com/apple'></iframe>
+    <iframe src='https://apple.com/apple'></iframe>
```
Doesn't this make two iframes share the process since they're both accessing https://apple.com?
We ideally have another test case or modify this one to have two distinct cross-site addresses.

szewai committed:
[Site Isolation] Add support for text manipulation in cross-site frames
rdar://150383885
https://bugs.webkit.org/show_bug.cgi?id=292339

Reviewed by Ryosuke Niwa.

To make text manipulation work with site isolation, UI process should send message to all frame processes and collect
replies from all of them. This patch also ensures web process inspects all local frames by replacing
traverseNextRendered() with traverseNext(), as traverseNextRendered() will stop early when encountering a remote frame.

API test: SiteIsolation.CompleteTextManipulation
          SiteIsolation.CompleteTextManipulationFailsInSomeFrame

* Source/WebCore/editing/TextManipulationController.cpp:
(WebCore::TextManipulationController::completeManipulation):
* Source/WebCore/editing/TextManipulationController.h:
* Source/WebCore/editing/TextManipulationControllerManipulationFailure.h:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _completeTextManipulation:completion:]):
(-[WKWebView _completeTextManipulationForItems:completion:]):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::startTextManipulations):
(WebKit::WebPageProxy::completeTextManipulation):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::startTextManipulations):
(WebKit::WebPage::completeTextManipulation):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(-[SiteIsolationTextManipulationDelegate init]):
(-[SiteIsolationTextManipulationDelegate _webView:didFindTextManipulationItems:]):
(-[SiteIsolationTextManipulationDelegate items]):
(TestWebKitAPI::createToken):
(TestWebKitAPI::createItem):
(TestWebKitAPI::TEST(SiteIsolation, CompleteTextManipulation)):
(TestWebKitAPI::(SiteIsolation, CompleteTextManipulationFailsInSomeFrame)):
(TestWebKitAPI::TEST(SiteIsolation, DISABLED_Events)): Deleted.

Canonical link: https://commits.webkit.org/294454@main
